Declare Insecure Redirects

DomT4 opened the pull request:
Re discussion in #38824.

vszakats commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -262,6 +262,10 @@ def initialize(name, resource)
   end
 
   def fetch
+    if @url.include?("https://downloads.sourceforge.net") || @url.include?("https://download.gnome") || @url.include?("https://www.apache.org/dyn/")
```
Wouldn't `https://download.gnome.org` be better than `https://download.gnome` here? `.com` and `.net` aren't used by the GNOME project, though I haven't checked other TLDs.

DomT4 replied:
I cut off the country code because I was trying to shorten an already lengthy line. ` Sourceforge `'s could also be cut for the same reason.

I don't mind either way though. We can do either, ultimately, if the maintainers like the idea.

vszakats replied:
No strong preference from here either. One interesting thing is that `sourceforge.jp` uses a different setup than the main sourceforge site, so that one is probably best to be handled separately.

vszakats commented:
This is good as a general disclaimer to avoid false expectations about security, my minor concern is that it will also appear when these redirectors start to switch using secure URLs, thus giving unnecessary worries for a safe redirect. Ideally they should switch _all_ their mirrors to SSL/TLS at once, but in practice most probably it will be a slow and gradual process (possibly with some plaintext mirrors remaining indefinitely).

DomT4 commented:
> thus giving unnecessary worries for an otherwise secure redirect.

This is largely why I softened the language from "will" and "won't" to "may" and "may not". I favour obeying the majority in this sort of thing; At the moment the majority (almost all, sadly) of the redirects are done to plaintext, so the warning is merited. Once we see the redirects hitting SSL/TLS more than plaintext, I'm happy to murder the warning, or narrow it down more than it is currently.

vszakats commented:
Thanks, noted and agreed.

jacknagel commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -262,6 +262,10 @@ def initialize(name, resource)
 
   def fetch
+    if @url.include?("https://downloads.sourceforge.net") || @url.include?("https://download.gnome") || @url.include?("https://www.apache.org/dyn/")
+      ohai "Please note this link may redirect to http and the download may not be protected by SSL/TLS."
```
s/link/URL/

DomT4 replied:
Changed :+1:

mikemcquaid commented:
> This is good as a general disclaimer to avoid false expectations about security, my minor concern is that it will also appear when these redirectors start to switch using secure URLs, thus giving unnecessary worries for an otherwise secure redirect.

Might be worth trying something where we use `curl` or similar to work out the actual redirected URL.

vszakats commented:
`-I -o /dev/null -w '%{url_effective}'` `curl` options, when appended to the normal download command should return the effective URL on stdout.

DomT4 commented:
I have some concerns about that being misleading myself. See comment on other thread. On phone atm but will dig out a link when not.

Sent from my iPhone

> On 20 Apr 2015, at 09:34, Mike McQuaid <notifications@github.com> wrote:
> 
> This is good as a general disclaimer to avoid false expectations about security, my minor concern is that it will also appear when these redirectors start to switch using secure URLs, thus giving unnecessary worries for an otherwise secure redirect.
> 
> Might be worth trying something where we use curl or similar to work out the actual redirected URL.
> 
> —
> Reply to this email directly or view it on GitHub.
>

vszakats commented:
I think I _may_ have understood your concern: If the redirection detection and the actual download are two separate `curl` sessions, the detected redirects _may_ (and if they may, they certainly will, as your tests has shown) result in two _different_ redirected URLs. To avoid that, it's possible to add the `-w` option to the actual download command, f.e.:
```
curl -sS -L 'http://sourceforge.net/projects/upx/files/upx/3.91/upx-3.91-src.tar.bz2'  -o 'dl.tar.bz2' -w '%{url_effective}'
```
Effective URL will be on stdout.

DomT4 commented:
Changing the ` curl ` args themselves is somewhat difficult, well, not difficult but more invasive - We spread the curl args out across multiple points - [global](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/global.rb#L93), [utils](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/utils.rb#L165-L179) and obviously, the [download_strategy](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/download_strategy.rb#L305-L307). Some of those elements are tagged as private API.

Feelings on changing any of that @jacknagel @mikemcquaid?

jacknagel commented:
Adding the argument wouldn't be too bad, but at first glance, it looks like changing how curl itself is invoked in order the capture the output would be difficult while maintaining compatibility.

One possibility would be something like this:

```diff
diff --git a/Library/Homebrew/download_strategy.rb b/Library/Homebrew/download_strategy.rb
index 589147c..d24a06e 100644
--- a/Library/Homebrew/download_strategy.rb
+++ b/Library/Homebrew/download_strategy.rb
@@ -313,6 +313,7 @@ def downloaded_size
   def curl(*args)
     args << '--connect-timeout' << '5' unless mirrors.empty?
     args << "--user" << meta.fetch(:user) if meta.key?(:user)
+    args << "-w" << "Downloaded from: %{url_effective}\n"
     super
   end
 end
```

but without capturing the output, this would print even in cases where no redirects occur.

I dunno.

vszakats commented:
Another alternative is to create an option to not allow protocol changes via redirections. This can be implemented by adding `--proto-redir =https` or `--proto-redir =http` `curl` option depending on the original URL being `https://` or `http://` respectively. The warning message in this patch could advertise such option.

[ In case the download would fail this way, Apache now has an implicit https mirror and similar can be implemented for sourceforge/gnome now or in the future. ]

mikemcquaid commented:
> Another alternative is to create an option to not allow protocol changes via redirections. This can be implemented by adding --proto-redir =https or --proto-redir =http curl option depending on the original URL being https:// or http:// respectively. The warning message in this patch could advertise such option.

This makes sense to me.

DomT4 commented:
> option depending on the original URL being https:// or http:// respectively. 

I put in a version of this. I don't mind at all if links want to silently upgrade themselves from plaintext to SSL/TLS - Breaking that side of things would send the wrong message. We're only interested in offering a way to block insecure redirects.

I've added an environmental variable to the mix here, along the lines of your suggestion. I've also changed the download message to mention that variable. Currently looks like this in practice:

![screen shot 2015-04-26 at 03 26 22](https://cloud.githubusercontent.com/assets/6998367/7335629/0f9fc0d6-ebc4-11e4-9e27-b18fd5b68223.png)

Thoughts welcome from everyone.

DomT4 commented:
If maintainers are happy with the above, I'll probably disable the ` this URL may redirect to http ` etc message if the environmental variable *is* set, since it won't be true in that case and it'll just noise up downloads without serving a purpose.

vszakats commented:
I like these. Only one part I'd argue: protocol upgrades. While convenient and slightly better than full plaintext, these can't IMO be considered secure, because in the initial (or any in-between) plaintext phase the redirection is very easy to spoof and the URL is also visible to any listeners. So, if we're to call the option 'secure downloads only', it'd seem better to exclude them. In such case the right solution is to swap the initial URL in the formula to use SSL/TLS instead.

DomT4 commented:
> In such case the right solution is to swap the initial URL in the formula to use SSL/TLS instead.

To be put it somewhat bluntly, I don't particularly think most users, even those running with the environmental variable, can be consistently relied upon to submit package updates to bump URLs to the correct protocol.

My worry there is that we'll end up blocking a lot of things, and instead of users submitting the changes most of the time, they'll just ` unset ` the variable for that download. I want to improve security, but at the same time, I don't want to make using *any* Homebrew environmental variable an inconvenience to users.

I don't think we can argue this two ways. We say we use ` https://downloads.sourceforge ` and similar because some SSL/TLS is better than complete plaintext - But if we ban plaintext to SSL/TLS redirects under the ` INSECURE ` env we're essentially reversing that and saying links should be explicitly plaintext all the way or explicitly SSL/TLS all the way, surely?

vszakats commented:
I agree not to bug users with this. This could rather be something verified in the `brew audit` phase to help it getting fixed by contributors.

mikemcquaid commented:
> This could rather be something verified in the brew audit phase to help it getting fixed by contributors.

This sounds interesting and we'd review a patch to do this.

DomT4 commented:
> This could rather be something verified in the brew audit phase to help it getting fixed by contributors.

Aye. Agreed. We should do it, but it should be nudged onto formulae authors rather than folks who just want to use Homebrew to deal with package management and not be fussed about the inner workings and such.

vszakats commented:
I've updated my prototype shell script with capability to catch this case and show a message accordingly:
https://gist.github.com/vszakats/663dc8ccf1d49b903f8e

[sorry for no direct patch - it's my Ruby skills]

mikemcquaid commented:
> I've updated my prototype shell script with capability to catch this case and show a message accordingly:
> https://gist.github.com/vszakats/663dc8ccf1d49b903f8e
> 
> [sorry for no direct patch - it's my Ruby skills]

@vszakats It'd be good if you feel passionate about these things if you can try and work on some Ruby. I'd never written any Ruby before Homebrew and started with formulae and then moved to core code, being a maintainer and now work full-time writing (mostly) Ruby. There are lots of resources and people who will help you.

vszakats commented:
Can't argue with that. I'll need to finish some running projects to allocate time for it. What would normally be a 5-60 minutes patch, ATM requires days, including to find out how to create/run a spare working copy of Homebrew where debugging is done. Ruby has a lots of ways to achieve the same result, which requires an extra leap to create something presentable. I'm hoping to get there eventually.

mikemcquaid commented:
Thanks @vszakats, much appreciated.

DomT4 commented:
Pushed a change so the message doesn't print if the environmental variable is set.

vszakats commented:
👍

DomT4 commented:
I'm happy enough with this if everyone else is at this point, but will adjust if there's any further feedback.

mikemcquaid commented:
Any other @Homebrew/owners have thoughts?

jacknagel commented:
Requesting a day to look it over, haven't kept up with the current proposal (sorry).

tdsmith commented on Library/Homebrew/manpages/brew.1.md:
```diff
@@ -556,6 +556,11 @@ can take several different forms:
+  * HOMEBREW\_NO\_INSECURE\_REDIRECT:
+    If set, Homebrew will not permit redirects from secure download protocols
+    to insecure ones. Whist ensuring your downloads are fully secure, this can
+    have a major impact on Sourceforge, Apache & GNOME based formulae.
```
Can this be more specific than "have a major impact"?

DomT4 replied:
I can be slightly more blunt I guess. `Will miserably break downloads of from-source Sourceforge, Apache & GNOME based formulae`? :wink:

DomT4 replied:
Tweaked the language a bit. Let me know if you're :smile: or :disappointed: with the new tweak.

tdsmith replied:
I like the new wording better, thanks!

tdsmith commented:
Can we do something subtler? This will get printed a lot and encourages users to do something that will break a lot of popular downloads. The messaging implies that these downloads aren't secure, which I think is confusing; information is leaked about which package you're downloading, but package integrity is still ensured. I don't know if this approach is right but maybe "private" is more useful than "secure."

DomT4 commented:
It won't print for anyone using a bottle, which softens the impact a decent chunk. The warning was less verbose in the first place, but this thread discussion developed it into something more thorough, heh.

We could be less pointy towards the ENV I guess, by suggesting people read the manpage to discover how to disable the plaintext redirects. Given most people's hatred of manpages, that could be a deterrent. I lean towards feeling like we've shoved enough dire, harsh warnings in to put off most people, but I could be underestimating folks' curiosity :smile_cat:.

I think "secure" is actually accurate here, sadly. "Private" doesn't make much sense in a way - Sourceforge still knows what you're downloading and keeps fairly extensive stats on that, regardless of SSL/TLS. I think ` https => http ` redirects is a valid security issue, even with our checksum usage.

vszakats commented:
> information is leaked about which package you're downloading, but package integrity is still ensured

I think package integrity cannot be ensured with plaintext URLs, because checksums are (I assume in most cases) calculated by formula authors of packages downloaded over the same insecure link. Homebrew users can only be sure that their downloaded copy (and the copy used to create bottles) is the same as the one the formula author used; not that it is an intact one.

Integrity could be ensured by using a signature, by getting the hash via a secure link, and/or by getting the content via a secure link.

tdsmith commented:
If the same checksum can be obtained from different geographical locations (minimally, bot and end user, in the event that a PR is submitted by an attacker) I'm not concerned about MITM attacks. If the origin server is compromised, HTTPS wouldn't have helped anyway and it doesn't make sense to suggest that HTTP transport is weaker. I don't think the package integrity exposure is significant enough to degrade the UX with a warning message like this about security.

Infrastructure separate from the build bot that made sure the build bot isn't being MITM'd by fetching links through tor and verifying checksums or something could be a helpful improvement but is out of scope of the current proposal.

vszakats commented:
In the current context there is no mechanism to ensure that a checksum is obtained from multiple locations (_if_ there are multiple locations). Even if there would be, it would only mean better probability that it's a valid one. Which, one can argue, is the same goal of SSL/TLS; but the latter is designed to do this, while the former relies on chance. I think it's unfair to call such system secure.

tdsmith commented on Library/Homebrew/manpages/brew.1.md:
```diff
@@ -556,6 +556,14 @@ can take several different forms:
+    If set, Homebrew will not permit redirects from secure download
+    protocols to insecure ones.
+
+    Whist ensuring your downloads are fully secure, this is likely
```
whilst or while, maybe?

DomT4 replied:
Seems to be a British/American lexicon difference thing. I'm happy to change it.

DomT4 replied:
Whoops. There's a typo there that I didn't spot, even after reading your comment. Apologies. Happy with "while" though, making the change now.

tdsmith commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -262,6 +262,16 @@ def initialize(name, resource)
+      puts <<-EOS.undent
+        This URL may redirect to http and the download may not be protected by SSL/TLS.
+
+        You can disable insecure redirect by setting the "HOMEBREW_NO_INSECURE_REDIRECT" environmental variable.
```
environment, not environmental

DomT4 replied:
Wasn't sure either way on this. The latter is used in a few places:

https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/doctor.rb#L793

https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/External-Commands.md#shell-scripts

DomT4 commented:
Tim has a valid point about checksum security - Although they are user generated the vast majority of the time, the test bots check the checksum over three times - Once for each OS X version. That checksum is then verified again on every user download.

Although not impossible, compromising each run through would be difficult given Sourceforge's regular switching of mirrors per connection. It would be a hell of an attack to compromise that many geographically variable connections.

I disagree with Tim on UX and the privacy/security language, and think this or something like this is worthwhile, but I'm pretty much with him on the checksum issue.

vszakats commented:
If you limit the context to Sourceforge, it may indeed give a good, practical level of assurance, knowing how it works now and have worked in the past, but it's not something you can automatically extend to other providers, or even to a future configuration of Sourceforge (or Apache, or Gnome).

I wish mirroring would move to SSL+TLS to avoid discussions like this.

tdsmith commented:
> I wish mirroring would move to SSL+TLS to avoid discussions like this.

I agree.

Is it valuable to have an option that blocks HTTPS to HTTP redirects but allows ordinary HTTP requests to succeed?

DomT4 commented:
> I wish mirroring would move to SSL/TLS to avoid discussions like this.

This is the promise land.

> Is it valuable to have an option that blocks HTTPS to HTTP redirects but allows ordinary HTTP requests to succeed?

The option in this PR should enforce that. If you want to make a ` http ` connection to Sourceforge, Apache or GNOME it won't flag any message at all. Unless I misunderstood you?

tdsmith commented:
Right, but if a user is paranoid enough to actually switch this on and suffer the ensuing breakage (and I'm still not convinced we should let them), do you think they'd ever want to use http at all?

DomT4 commented:
I don't think the issue at hand is insecure links per se - The issue is insecure links *appearing* to be secure links. We only show the `https://downloads.sourceforge ` etc part - We don't show the part where it redirects to plaintext.

As far as most users are concerned, that download is handled by a secure protocol. Almost nobody is going to bother to follow the redirect and poke around the mirroring system themselves. Jack's (perhaps semi-throwaway) comment that started this rolling was a worry/frustration(?) that we were misleading users.

Extremely broadly speaking people, through the social training of browsers, have come to associate ` https ` with protected and ` http ` as unprotected; showing them one thing that looks secure whilst doing another that is less secure in the background silently is problematic. It's less of a problem with Apache because we print the Apache mirror being used, I believe.

FWIW, I'm not strongly against the idea of supporting an extremely optional total ban of plaintext download links, but I think it's a different issue to the one being discussed with this set of changes.

mikemcquaid commented:
> Extremely broadly speaking people, through the social training of browsers, have come to associate https with protected and http as unprotected; showing them one thing that looks secure whilst doing another that is less secure in the background silently is problematic. It's less of a problem with Apache because we print the Apache mirror being used, I believe.

Yes, I agree with this. I think any blocking (even redirect blocking) should be optional and otherwise if we can just print the redirect location that'd be good. It would probably help with debugging other issues too.

Really the people to work with here are SourceForge and friends to get more SSL secured mirrors and deal with the issues here.

tdsmith commented:
This is particularly important to browsers because people make decisions about entering their credit card number or their password based on looking at a URL, which isn't the case here. I'd like to show the final URLs from which resources are fetched. I don't think verbose warnings or options to block redirects are really useful.

A simple hack would be to add
`"-w", "Fetched %{url_effective} in %{time_total} s.\\n"`
which always displays the actual URL curl fetches (and also shows the elapsed time, which makes us feel better about always printing this even if the URL doesn't contain new information).

DomT4 commented:
> This is particularly important to browsers because people make decisions about entering their credit card number or their password based on looking at a URL, which isn't the case here.

If it were as black-and-white as this, there would be no justification for any download server to serve anything over secure channels. Encryption has a broader use and [push](http://www.wired.com/2014/04/https/) beside just protecting the obvious. Both the IAB and IETF have decried plaintext connections inside the last year, re the IETF:

> User trust is critical to the Internet's continued growth and evolution. Realizing the IAB's aspiration would drastically reduce the ability to eavesdrop or modify information sent over the Internet.

Between this thread, the last one, and the one over at the Caskroom there's been enough conflicting ideas that there's a fairly decent chance we all throw around different ideas and implementations for months whilst the issue persists. I just tried to land on a middle ground here :neutral_face:.

Sadly, re Jack's suggestion in [this comment](https://github.com/Homebrew/homebrew/pull/38835#issuecomment-95772259) we do break *at least* Apache downloads. The reasonable thing to do consequently is scope it, but that means we won't see the actual download URL for every formulae, so it's not a huge debug gain.

It also doesn't give the user any actual *choice* in the matter - Before they are notified where the download is coming from, it has already happened; If you didn't want a plaintext download, tough, it's finished. *"Hey this isn't a secure link, and we're gonna tell you about it, but you're stuck with it"* isn't great UX either.

```diff
diff --git a/Library/Homebrew/download_strategy.rb b/Library/Homebrew/download_strategy.rb
index d270ec4..86be24d 100644
--- a/Library/Homebrew/download_strategy.rb
+++ b/Library/Homebrew/download_strategy.rb
@@ -317,8 +317,11 @@ class CurlDownloadStrategy < AbstractFileDownloadStrategy
   end

   def curl(*args)
+    insecure_redirects = @url.include?("https://downloads.sourceforge.net") || @url.include?("https://download.gnome")
+
     args << '--connect-timeout' << '5' unless mirrors.empty?
     args << "--user" << meta.fetch(:user) if meta.key?(:user)
+    args << "-w" << "Downloaded from: %{url_effective}\n" if insecure_redirects
     super
   end
 end
```

Whatever we do, we're not going to hit a perfect solution because we're forced to work with the hand upstream has dealt us. We just have to decide on the least-sucky, most-effective, least-breaky solution.

tdsmith commented:
I have always agreed that encryption is good! I only meant that displaying a correct indication of whether a page is served from a secure origin is much more important for a browser than it is for Homebrew.

Just realized Jack already suggested the same thing I did; I had skimmed the top of the thread, oops. Anyway, I like it. :p I don't see the need to scope it to only a couple of hosts; I think the behavior is probably more common than that.

DomT4 commented:
> I don't see the need to scope it to only a couple of hosts; I think the behavior is probably more common than that.

I agree, but we have a couple of problems.

* It breaks Apache downloads and forces mirror fallback. It's hard to guarantee that breakage won't seep into any other bracket of formulae without fairly deep testing.
* It looks.... busy when using ` brew fetch ` on formulae with patches:

![screen shot 2015-04-29 at 04 43 49](https://cloud.githubusercontent.com/assets/6998367/7384404/5e232cbc-ee2a-11e4-9447-9d769efa67b9.png)

mikemcquaid commented:
I think it would be an idea to get the actual URL before downloading and print it only if it differs to what we have set as the `url`. Could do this by doing a `HEAD` request with `curl` and following the redirects or something?

vszakats commented:
@mikemcquaid I like this, reading the redirect, displaying it if it differs and unsafe, stop if user opted so, otherwise continue to download the effective URL directly (this time without `-L` option).

mikemcquaid commented:
I think it'd be worth displaying it even if "safe" but just if different to the URL. Would fix another class of issues we occasionally see where a random proxy/cafe wifi redirects to a weird URL.

mikemcquaid commented:
but the option could stop then if redirecting to "unsafe", yeh.

vszakats commented:
> I think it'd be worth displaying it even if "safe" but just if different to the URL.

Agreed.

DomT4 commented:
We're happy keeping the blocking option? Will throw a couple more hours into chasing this around tonight.

Sent from my iPhone

> On 29 Apr 2015, at 15:21, Mike McQuaid <notifications@github.com> wrote:
> 
> but the option could stop then if redirecting to "unsafe", yeh.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mikemcquaid commented:
Thanks @DomT4! I reckon I'd do the logic like this:

1. Do a HEAD request to see what the redirected URL is going to be and get it into a Ruby variable
2. Compare that to the actual URL
3. If the option is set and e.g. it's changing from HTTPS to HTTP: error out
4. If the redirected URL is different from the intended URL: print it instead

vszakats commented:
> We're happy keeping the blocking option?

I think it's useful. Those caring about this could let the download fall back to the formula `mirror`, which may offer a proper SSL+TLS URL (and without unnecessarily overloading those mirrors).

DomT4 commented:
> I think it's useful.

I agree, I'd *like* to keep it. I just didn't know how the maintainers felt about it given Tim's points.

> I reckon I'd do the logic like this:

Makes sense to me. I threw around some basic implementations last night of something very similar, but wasn't hugely happy with them. I'll do some more on it tonight and try not to be so grumpy about my own code :smile_cat:.

vszakats commented:
Thanks a lot @DomT4

DomT4 commented:
Alright, I've pushed a new implementation. Some notes:

* I added a broad rescue but this rescue only kicks in after every other possibility is exhausted. It just ensures that if there's no "Location" field in the ` HEAD ` request things don't blow up. (Pretty much everything that didn't have a "Location" field was blowing up otherwise. We use it elsewhere so I don't *think* my usage here is too heinous).
* I still have so much Ruby to learn; the code used is probably not the tidiest in the world, or as simplistic as it could be. Feel free to tear it apart - I learn more when you're brutal, much to my dismay :wink:.

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,23 @@ def initialize(name, resource)
+      does_redirect
+
+      if !does_redirect.include?(@url) && ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
+        ohai "Downloaded from #{does_redirect}".gsub("Location: ", "")
```
Has this already done the redirect or should it be in present tense?

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,23 @@ def initialize(name, resource)
+        ohai "Downloaded from #{does_redirect}".gsub("Location: ", "")
+      elsif @url.include?("https") && does_redirect.include?("http")
+        puts "Insecure redirect detected! Escaping to safety."
+        exit! 1
```
Probably wants to be an exception (see others in this file) rather than an `exit`

DomT4 replied:
I managed to screw myself into forcing an ` exit ` here by doing ` rescue Exception => e ` further down :smile_cat:. Suppose by narrowing down the rescue we can make this an exception.

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,23 @@ def initialize(name, resource)
+    # Don't care if it fails at this point. If it fails at this point the primary
+    # reason is that does_redirect did not return a "Location" field, i.e. the link
+    # does not silently redirect elsewhere, which is good.
+    rescue Exception => e
```
Probably a bit broad.

DomT4 replied:
` rescue Exception ` any less broad? Wasn't sure what the best way to narrow this down was. Essentially anything that fails on the above two blocks needs to be saved from blowing everything up, heh.

xu-cheng replied:
If you use `reslove_url` I mentioned below, there shouldn't be any exception to be caught.

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -312,6 +329,10 @@ def _fetch
     curl @url, "-C", downloaded_size, "-o", temporary_path
   end
 
+  def does_redirect
```
Might want to make this `does_redirect?` and return a true or false value based on the `@url`

DomT4 replied:
Nice. Will do :+1:

xu-cheng replied:
Consider the usage above, I think it's better to make method similar to this:
```ruby
def resolve_url(url):
  m = Utils.popen_read("curl", "-I", url).match /^Location: (.+)$/
  m.nil? ? url : resolve_url(m.captures.first)
end
```

xu-cheng replied:
Also, if possible use `open-uri` rather than shellout would be better.

DomT4 replied:
I'm not overly familiar with ` open-uri `, and given we're already shelling out plenty in ` fetch ` I didn't think one more was particularly gratuitous. Also somewhat worry about Ruby not being great at backwards compatibility - Will ` open-uri ` work out of the box right back to 10.4/10.5's Ruby? I'd prefer not to create more work for Misty if not.

xu-cheng replied:
See my update.

mikemcquaid replied:
I think shelling out would be better here.

DomT4 replied:
Not sure what I'm screwing up with your contribution Xu. I stripped all of my code out just to check and it's still throwing failure.

```ruby

  def fetch
    ohai "Downloading #{@url}"

    begin
      resolve_url
    end

  def resolve_url(url)
    m = Utils.popen_read("curl", "-I", url).match /^Location: (.+)$/
    m.nil? ? url : resolve_url(m.captures.first)
  end
```

Failure is: ` Error: wrong number of arguments (0 for 1) `.

xu-cheng replied:
I mean you should do things like this:
```ruby
url = resolve_url(@url)
if url != @url
  if @url.start_with?("https://") && url.start_with?("http://") &&
     ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
    ofail "Insecure redirect detected! Escaping to safety."
  end
  ohai "Download from #{url}"
end
```

DomT4 replied:
Ah, of course. Duh, sorry. Thanks - Pushed.

xu-cheng commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,23 @@ def initialize(name, resource)
     ohai "Downloading #{@url}"
+
+    begin
+      does_redirect
```
Let's call this method only once.

DomT4 commented:
Code changed. Same purpose and end result as before, just without the need for exception rescue, and generally a bit easier for someone who isn't inside my mind to get the thinking behind everything (hopefully). CC @Homebrew/owners for review/comment, hopefully for the last time :smile_cat:.

xu-cheng commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,17 @@ def initialize(name, resource)
+         !ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
+         raise "Insecure redirect detected! Escaping to safety."
+      else
+        ohai "Downloading from: #{url}".gsub("Location: ", "")
```
You don't need the `gsub` part.

DomT4 replied:
Sorry, lazy copy/paste from previous block. Fixed.

vszakats commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -312,6 +323,11 @@ def _fetch
 
+  def resolve_url(url)
+    m = Utils.popen_read("curl", "-I", url).match /^Location: (.+)$/
+    m.nil? ? url : resolve_url(m.captures.first)
```
Isn't it a concern that a(n https -> http) redirect could happen in a second or any subsequent redirect?

DomT4 replied:
Sure, if we can find a reproducible example. I guess we could reintroduce ` --proto-redir ` but I'm not huge on introducing extra code unless there's a provable use case for it. A link that darts from secure => secure => insecure is a relative rarity.

vszakats replied:
The issue is we can't know if and when such redirect would occur; it completely depends on the providers' actual configuration (f.e. as the they are gradually upgrading the internal elements of a redirect chain to SSL/TLS, or a secure mirror may be (mis)configured to redirect to a backup server without SSL/TLS), regardless of what the current configuration is. My vote for `--proto-redir`, to be on the secure side. But, even with `--proto-redir`, why manually parsing the headers of the first redirect is preferred over `-w '%{url_effective}'`?

mikemcquaid replied:
If we add `-L` to the arguments here and get the last `Location:` it looks like that'll do the above without needing `proto-redir`.

xu-cheng replied:
A quick note here. It should be `m.nil? ? url : URI.join(url, resolve_url(m.captures.first)).to_s`

xu-cheng commented:
After reviewing all of old comments, I think we can do this:
* Moving `resolve_url` to `utils.rb` with following implementations:
```ruby
def resolve_url(url, &block)
  m = Utils.popen_read("curl", "-I", url).match /^Location: (.+)$/
  if m.nil?
    url
  else
    url = m.captures.first.chomp
    yield url if block_given?
    resolve_url(url)
  end
end
```
* Add `ARGV.no_insecure_redirect?` method (maybe a shorter name):
```ruby
def no_insecure_redirect?
  include?("--no-insecure-redirect") || !ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
end
```
* Change `CurlDownloadStrategy#_fetch` like this:
```ruby
raise_when_insecure = ARGV.no_insecure_redirect? && @url.start_with?("https://")
verbose = ARGV.verbose?

final_url = resolve_url(@url) do |url|
  if raise_when_insecure && !url.start_with?("https://")
    raise "Insecure redirect detected! Escaping to safety."
  end

  ohai "Downloading from: #{url}" if verbose
end

if final_url != @url && !verbose
  ohai "Downloading from: #{url}"
end

@url = final_url
```
* Using `resolve_url` in `audit.rb` in similar fashion.

vszakats commented:
The latest version seems to ignore the initial concern of @DomT4, that subsequent requests may result in different redirected URLs (some potentially https, some not and being potentially different from the actual effective download URL used in the subsequent download request, f.e. `sf.net`). At the same time it drops the initial suggestions to use `-w` and `--proto-redir` options to detect and optionally ensure secure redirects, over a manual approach which doesn't handle multiple redirects and depends on manually parsing headers of the HTTP protocol. (both solutions rely on catching stdout/err). Is it the goal to not rely on any special `curl` options to be as tool-agnostic as possible?

xu-cheng commented:
> that subsequent requests may result in different redirected URLs

>  which doesn't handle multiple redirects

I think I have handle these situations.

> At the same time it drops the initial suggestions to use -w and --proto-redir options

I have no objection on `--proto-redir`. However it wouldn't print the final url for users who don't choose to escape from insecure redirect. `-w` would be too noise.  Manually parsing just gives us more custom on what to display.

vszakats commented:
> I think I have handle these situations.

Oh, I see the recursion now.

> `-w` would be too noise

Maybe this version would resolve the noise?:
```ruby
def resolve_url(url, &block)
  url = URI.parse(url) unless url.is_a? URI
  url = Utils.popen_read("curl", "-L", "-I", "-o", "/dev/null", "-w", "%{url_effective}", url)
end
```

One outstanding issue (with both versions) is with a redirect chain having `http` step(s) in the middle:
`https` &rarr; `http` &rarr; `https`
This would currently pass, while it shouldn't.

vszakats commented:
Minor note on this line:
```ruby
m = Utils.popen_read("curl", "-I", url.to_s).match /^Location: (.+)$/
```

It badly interferes with a `.curlrc` that contains [`--remote-name-all`](https://changelog.com/magic-curl-features-with-daniel-stenberg/), redirecting the output to a disk file instead of stdout. Solution is to always specify a `-o` option when calling `curl`.

DomT4 commented:
> It badly interferes with a .curlrc containing

I've always wondered why we're not just passing `-q` to our `curl` arguments given we have strict and harsh warnings in place in `brew doctor` about `.curlrc` existing. CC @jacknagel?

jacknagel commented:
We used to, people complained.

DomT4 commented:
Ah, frustrating. Does anyone have any objections to Xu's rewrite? Or thoughts on the `proto-redir ` situation?

I'll rewrite it later tonight, but somewhat concerned three weeks into this PR we've almost completely rewritten it to serve an ever expanding purpose ~4 times and don't seem any closer to agreement on acceptable compromises/end purpose.

xu-cheng commented:
> I've always wondered why we're not just passing -q to our curl arguments given we have harsh warnings in place in brew doctor about .curlrc existing. 

`.curlrc` can be important in some user cases such as proxy.

> I'll rewrite it later tonight, but somewhat concerned three weeks into this PR we've almost completely rewritten it to serve an ever expanding purpose ~4 times and don't seem any closer to agreement on acceptable compromises/end purpose.

In fact, I'm not that confidence on my implementation. There are some edge cases which are easy to ignored. For example, I failed to use `URI.join` in the first time. I do have a question on whether we should print the effective url for normal users(who don't set  `HOMEBREW_NO_INSECURE_REDIRECT `)? If not, `proto-redir` may be a more robust solution.

vszakats commented:
I think the overall goal is to help the shift towards secure communication. One component of this may be to make it visible if insecure things are going on, another is to prevent a protocol downgrade - or even better _any protocol change_. But there could be more; f.e. to prioritize secure download sources if such is available (i.e. when SSL/TLS URL is specified as a `mirror`), or to prevent using insecure protocols altogether. Next is how to make this user controllable: using granular options, one switch, or maybe levels.

I'd personally opt for levels; these three of them:
- default - current, with warning on protocol downgrade (can be extended with more in the future)
- prefer-secure - try using secure mirrors first, disallow protocol change, warn when falling back to an insecure download
- only-secure - disallow anything insecure (ftp, http, git, cvs, svn)

my 2cents. Ignore if it doesn't make sense or veers too far off this discussion.

tdsmith commented:
I think limiting the scope of this change to "try not to lie to the user" is cleanest. Adding a `proto-redir` setting will break things; maybe setting that in curlrc is a good option for people who want it.

vszakats commented:
You cannot set it in .curlrc. You need to use different options for http URLs and https URLs.

vszakats commented:
I opened feature requests in `curl` for that a while ago, but it is yet to be accepted: https://github.com/bagder/curl/issues/226

Anyhow `.curlrc` is a pretty bad solution for this problem, because it will apply for _all_ `curl` operation, and this is rarely viable in real life.

DomT4 commented:
> I do have a question on whether we should print the effective url for normal users

I am very strongly in favour of printing the effective URL regardless of what settings the user has enabled. As Mike highlighted further up, It helps us find cases which redirect which we're not currently aware of and it can be useful in terms of dealing with bug reports. It's also honest and transparent to users; even if they choose not to do anything about it, they'll know which protocol their download is getting served over.

My point of view is that we make users aware of what's going on, or rather what was already silently happening, and then if they want to mitigate that we have an option available for them to do so.

> But there could be more; f.e. to prioritize secure download sources if such is available (i.e. when SSL/TLS URL is specified as a mirror)

All the GNU stuff, all the Apache stuff, both have firm requests for people to use the geolocation mirrors. If Homebrew starts hitting their core servers hard we may rub people the wrong way.

Before we offered something as drastic as that it'd probably be courteous to initiate some discussion with the various parties that we'd be likely to redirect the most traffic to. I think that is probably a project for a future PR; If we bury that in this PR we won't get it merged for months.

mikemcquaid commented:
> In fact, I'm not that confident on my implementation. There are some edge cases which are easy to be ignored. For example, I failed to use URI.join in the first time. I do have a question on whether we should print the effective url for normal users(who don't set HOMEBREW_NO_INSECURE_REDIRECT)? If not, proto-redir may be a more robust solution.

I'd like us to print this URL as I think it'll help with debugging and it's better to keep users informed if there's a protocol downgrade.

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,17 @@ def initialize(name, resource)
   def fetch
     ohai "Downloading #{@url}"
+
+    url = resolve_url(@url)
```
Might want to call this `resolved_url` or `actual_url` or something to make the below code a bit easier to follow

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,17 @@ def initialize(name, resource)
+    url = resolve_url(@url)
+    if url != @url
+      if @url.start_with?("https://") && url.start_with?("http://") &&
+         !ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
```
I'd maybe tweak the indentation here so it's more obvious it's grouped with the previous line.

mikemcquaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -269,6 +269,17 @@ def initialize(name, resource)
+    if url != @url
+      if @url.start_with?("https://") && url.start_with?("http://") &&
+         !ENV["HOMEBREW_NO_INSECURE_REDIRECT"].nil?
+         raise "Insecure redirect detected! Escaping to safety."
```
`https to http redirect detected and HOMEBREW_NO_INSECURE_REDIRECT is set so aborting`

mikemcquaid commented on Library/Homebrew/manpages/brew.1.md:
```diff
@@ -574,6 +574,14 @@ can take several different forms:
 
+  * HOMEBREW\_NO\_INSECURE\_REDIRECT:
+    If set, Homebrew will not permit redirects from secure download
+    protocols to insecure ones.
```
Might want to just say HTTPS/HTTP here as it'd just handling that and not e.g. FTP.

mikemcquaid commented:
To me the diff looks pretty much what I had in mind (with a few tweaks).

DomT4 commented:
Changes pushed in response to comments. More comment welcome :wink:

xu-cheng commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -312,6 +323,11 @@ def _fetch
   end
 
+  def actual_url(url)
+    m = Utils.popen_read("curl", "-I", "-L", url).match /^Location: (.+)$/
```
This won't work. We cannot add `-L`. If you test it, you will see why.

DomT4 replied:
Yes that's what I thought to but I did test it, it worked. Have a reproducible example?

Sent from my iPhone

> On 16 May 2015, at 17:28, Xu Cheng <notifications@github.com> wrote:
> 
> In Library/Homebrew/download_strategy.rb:
> 
> > @@ -312,6 +323,11 @@ def _fetch
> >      curl @url, "-C", downloaded_size, "-o", temporary_path
> >    end
> >  
> > +  def actual_url(url)
> > +    m = Utils.popen_read("curl", "-I", "-L", url).match /^Location: (.+)$/
> This won't work. We cannot add -L. If you test it, you will see why.
> 
> —
> Reply to this email directly or view it on GitHub.
>

DomT4 replied:
It slows the GNU formulae `fetch` process down slightly. That's the only problem I've seen though. I `fetch`ed several SF formulae, several known non-redirected formulae, bottles and source packages. What problem are you expecting to run into here?

xu-cheng replied:
I didn't test it on a resource but on my own website.

```
$ brew pry
==> Interactive Homebrew Shell
Example commands available with: brew pry --examples
[1] brew(main)> def actual_url(url)
[1] brew(main)*   m = Utils.popen_read("curl", "-I", "-L", url).match /^Location: (.+)$/
[1] brew(main)*   m.nil? ? url : URI.join(url, actual_url(m.captures.first)).to_s
[1] brew(main)* end
=> :actual_url
[2] brew(main)> actual_url "http://www.xuc.me"
URI::InvalidURIError: bad URI(is not URI?): /
from /usr/local/Cellar/ruby/2.2.2/lib/ruby/2.2.0/uri/generic.rb:1100:in `rescue in merge'
[3] brew(main)>

$ curl -I -L http://www.xuc.me
HTTP/1.1 302 Found
Date: Sat, 16 May 2015 17:09:17 GMT
Connection: keep-alive
Set-Cookie: __cfduid=def8c22540ebda317655e556fef3699ed1431796157; expires=Sun, 15-May-16 17:09:17 GMT; path=/; domain=.xuc.me; HttpOnly
Pragma: no-cache
cache-control: no-cache
Location: /
Server: cloudflare-nginx
CF-RAY: 1e78b07e9429198c-HKG

HTTP/1.1 301 Moved Permanently
Date: Sat, 16 May 2015 17:09:18 GMT
Content-Type: text/html
Connection: keep-alive
Set-Cookie: __cfduid=def8c22540ebda317655e556fef3699ed1431796157; expires=Sun, 15-May-16 17:09:17 GMT; path=/; domain=.xuc.me; HttpOnly
Location: http://xuc.me/
Server: cloudflare-nginx
CF-RAY: 1e78b08194a1198c-HKG

HTTP/1.1 301 Moved Permanently
Date: Sat, 16 May 2015 17:09:18 GMT
Connection: keep-alive
Set-Cookie: __cfduid=ddf971c5bcd00711d97cebfc66f78c9221431796158; expires=Sun, 15-May-16 17:09:18 GMT; path=/; domain=.xuc.me; HttpOnly
Location: https://xuc.me/
Server: cloudflare-nginx
CF-RAY: 1e78b084d9a200c4-HKG

HTTP/1.1 200 OK
Server: cloudflare-nginx
Date: Sat, 16 May 2015 17:09:18 GMT
Content-Type: text/html; charset=utf-8
Connection: keep-alive
Set-Cookie: __cfduid=dc3f4faf4b1bb38c89b7f6c00ec55dff51431796158; expires=Sun, 15-May-16 17:09:18 GMT; path=/; domain=.xuc.me; HttpOnly
Last-Modified: Mon, 16 Mar 2015 01:33:56 GMT
Access-Control-Allow-Origin: *
Expires: Sat, 16 May 2015 17:19:18 GMT
Cache-Control: max-age=600
CF-RAY: 1e78b0863c6b1944-HKG
```

DomT4 replied:
For me:

```
==> Interactive Homebrew Shell
Example commands available with: brew irb --examples
irb(main):001:0> def actual_url(url)
irb(main):002:1>   m = Utils.popen_read("curl", "-I", "-L", url).match /^Location: (.+)$/
irb(main):003:1>   m.nil? ? url : URI.join(url, actual_url(m.captures.first)).to_s
irb(main):004:1> end
=> nil
irb(main):005:0> actual_url "http://www.xuc.me"
=> "http://www.xuc.me/"
irb(main):006:0>
```

Tested using both Ruby 2.2.2 via `rbenv` and using Ruby 2.2.2 from Homebrew.

@mikemcquaid @jacknagel @vszakats Can you run the above as well and see what reproduces for you?

xu-cheng replied:
Nevertheless, `curl -I -L` will print all redirections while we only catch one `Location:`. And it is contradicted to the recursive in the code.

xu-cheng replied:
If you want to use `-I -L`, it should be something like this:
```ruby
def actual_urls(url)
  Utils.popen_read("curl", "-I", "-L", url).scan(/^Location: (.+)$/).reduce([url]) do |urls, location|
    urls << URI.join(urls.last, location.first.chomp).to_s
  end
end
```

vszakats replied:
I'm having an error `bad URI(is not URI?): http://sourceforge.net/projects/upx/files/upx/3.91/upx-3.91-src.tar.bz2`, though this is most probably due to something I hosed while creating an isolated test snippet.

In the meanwhile, this version works:
```ruby
def actual_url(url, &block)
  Utils.popen_read("curl", "-L", "-I", "-o", "/dev/null", "-w", "%{url_effective}", url)
end
```

DomT4 replied:
@xu-cheng Your version causes some issues:

```
==> Downloading https://downloads.sourceforge.net/project/ccd2iso/ccd2iso/ccd2iso-0.3/ccd2iso-0.3.tar.gz
Error: undefined method `start_with?' for #<Array:0x007fa98a94f5f8>
```

I adjusted the syntax at the top to fit the new `actual_urls` language but it still grumped at me. @vszakats' version worked if people are happy with that?

DomT4 replied:
@xu-cheng ^ Re your comment in the main thread, I hit issues when I attempted to use your syntax?

xu-cheng replied:
That's because `actual_urls` return a list of urls which you should check all of them.

DomT4 replied:
I attempted to make some adjustments when you proposed it but hit errors fairly consistently. What's the best way to handle this?

mikemcquaid replied:
@DomT4 You want:
```ruby
curl_out = Utils.popen_read("curl", "-I", "-L", url).lines
location_matches = curl_out.map {|l| l.match /^Location: (.+)$/}.compact
m = location_matches.last
m.nil? ? url : URI.join(url, actual_url(m.captures.first)).to_s
```

xu-cheng replied:
@mikemcquaid Sorry, but your version is still incorrect. `-L` cannot be used with code recursion. 

I implement my own version which can be viewed from https://github.com/Homebrew/homebrew/compare/Homebrew:master...xu-cheng:curl

mikemcquaid replied:
@xu-cheng Can you explain what's incorrect about mine? I tested it and it produced the expected result so I'm confused.

mikemcquaid replied:
Ah yes, I see the recursion issue.

xu-cheng replied:
Yes, the test may create false positive as there aren't many resources with multiple redirects.

DomT4 replied:
Which version am I using here? Your's or Xu's?

mikemcquaid replied:
Xu's.

smarek commented:
I've addressed this issue to GNOME sysadmins through this ticket, https://bugzilla.gnome.org/show_bug.cgi?id=749481

Which would be far more elegant solution, to get redirect only to valid HTTPS mirrors if the original file is accessed through HTTPS (which means, client is already able to verify SSL certificate chain or deliberately ignores it)

mikemcquaid commented:
I'd like to help test this so we can get it merged in. Can you suggest some formulae I can try to reproduce the insecure redirect (ideally one that has a single redirect and one that has more than one)? Thanks!

xu-cheng commented:
Sorry for the late comment. But current implementation is still incorrect. In short, you cannot use `-L` (which will print all the redirects) and recursion at the same time. 

My version of `actual_urls` is correct. Except it will output all redirect urls as an array.

DomT4 commented:
I need to deal with the multi-hop situation still - I'll rebase this and push that change ASAP.

> Can you suggest some formulae I can try to reproduce the insecure redirect

```
brew fetch bash -fs -v
brew fetch diction -fs -v
brew fetch cogl -fs -v
```

Those are all pretty small formulae that can (always in GNU's case) insecure redirect off the primary url. Any sourceforge formula will have the same issue, but SF downloads seem to be having issues at the moment (Both in browser and via wget/curl).

mikemcquaid commented:
@DomT4 the current output looks great! None of them seem to be `https` -> `http`, though.

DomT4 commented:
Thanks! GNOME's rejiggle of redirection priority helped a lot on that - Since they bumped the priority of SSL/TLS mirrors I've barely hit a plaintext one.

SF is the main redirector:

```
==> Downloading https://downloads.sourceforge.net/doublecpp/doublecpp-0.6.3.tar.gz
==> Downloading from: http://downloads.sourceforge.net/project/doublecpp/doublecpp/0.6.3/doublecpp-0.6.3.tar.gz
```

```
==> Downloading https://downloads.sourceforge.net/project/mussh/mussh/1.0/mussh-1.0.tgz
==> Downloading from: http://freefr.dl.sourceforge.net/project/mussh/mussh/1.0/mussh-1.0.tgz
```

```
==> Downloading https://downloads.sourceforge.net/project/arss/arss/0.2.3/arss-0.2.3-src.tar.gz
==> Downloading from: http://heanet.dl.sourceforge.net/project/arss/arss/0.2.3/arss-0.2.3-src.tar.gz
```

vszakats commented:
Another `https` &rarr; `http` redirector is Apache, f.e. formula `ant`.

mikemcquaid commented:
Thanks folks. Will test again when you've pushed the code changes (and I guess it should be good to merge then?). Shout when that's happened!

DomT4 commented:
Alright, code updated, manpage updated.

@Homebrew/owners @bfontaine - Let's give this a (hopefully) final review and :+1: / :-1: this.

DomT4 commented:
As an aside, and something I would prefer not to get super hung up on in this PR, Apache formulae *are* currently excluded from being impacted by the variable. This is explicit in the manpage.

I'd like to automatically shunt Apache formulae onto the secure/private `https` mirror if the variable is set in future, but before doing so I think it'd be polite to get in touch with Apache and ask them if they mind the potential extra traffic, or like GNOME if they could bump the priority of secure mirrors so we hit them far less often.

Given the vast numbers of Sourceforge (In particular) formulae in the core, and the general benefits of redirection being announced to both us and users, I think this is worth getting in beforehand.

mikemcquaid commented:
Why is Apache excluded, just the different download strategy? If it's not a lot of work it'd be good to enable the same for them for consistency but no worries if that's a later PR. I'll review this more in the next few days and merge if it looks good

DomT4 commented:
>  just the different download strategy?

This is a part of it, yeah. It doesn't share the same code as the standard `curl` strategy.

Beyond that though there's *always* a safe assumption that Apache formulae will have a secure mirror available, which we can use automatically if the `ENV` is set, once we talk to Apache upstream - Sadly we can't make the same assumption about any other group of formula, aside GNU, and GNU redirect from HTTP to HTTP anyway.

mikemcquaid commented:
This is great work here @DomT4, thanks for all your patience here and for your help @vszakats. Merged!

vszakats commented:
Thanks to all of you folks, great to see this finalized.

DomT4 commented:
Thanks for the feedback, assistance and patience in getting this done all :tada:.
